### ISSUE TITLE
Enable support for a single entity connection string

### ISSUE THREAD

**Issue by atyourservers** [paolosalvatori/ServiceBusExplorer#375]
I have a test Service Bus deployed with only one queue. Connecting to only the queue entity using Service Bus Explorer 4.1.113 works without any issues. Connecting to the queue and say the topic entities, generates the error shown below. This issue occurs with any entity combination selected when connecting to a queue, topic, et cetera directly.

```
<09:41:29> The application is now connected to the sb://some-service-bus-name.servicebus.windows.net/ service bus namespace.
<09:41:29> MessagingFactory successfully created
<09:41:29> The queue some-queue-name has been successfully retrieved.
<09:41:29> Failed to retrieve Service Bus topics. Exception: Microsoft.ServiceBus.Messaging.MessagingException: Cannot get entity 'some-queue-name' because it is not of type TopicDescription. Check that you are using method(s) with the correct entity type. ---> System.Runtime.Serialization.SerializationException: Error in line 1 position 2641. Expecting element 'TopicDescription' from namespace 'http://schemas.microsoft.com/netservices/2010/10/servicebus/connect'.. Encountered 'None'  with name '', namespace ''. 

              at System.Runtime.Serialization.DataContractSerializer.InternalReadObject(XmlReaderDelegator xmlReader, Boolean verifyObjectName, DataContractResolver dataContractResolver)

              at System.Runtime.Serialization.XmlObjectSerializer.ReadObjectHandleExceptions(XmlReaderDelegator reader, Boolean verifyObjectName, DataContractResolver dataContractResolver)

              at Microsoft.ServiceBus.Messaging.MessagingDescriptionSerializer`1.DeserializeFromAtomFeed(Stream stream)

              at Microsoft.ServiceBus.Messaging.ServiceBusResourceOperations.GetAsyncResult`1.<GetAsyncSteps>d__22.MoveNext()

              --- End of inner exception stack trace ---

              at Microsoft.Azure.ServiceBusExplorer.ServiceBusHelper.GetTopicUsingEntityPath() in C:\projects\servicebusexplorer\src\ServiceBusExplorer\Helpers\ServiceBusHelper.cs:line 2087

              at Microsoft.Azure.ServiceBusExplorer.ServiceBusHelper.GetTopics(String filter) in C:\projects\servicebusexplorer\src\ServiceBusExplorer\Helpers\ServiceBusHelper.cs:line 2062

              at Microsoft.Azure.ServiceBusExplorer.Forms.MainForm.GetEntities(EntityType entityType) in C:\projects\servicebusexplorer\src\ServiceBusExplorer\Forms\MainForm.cs:line 4477
```

The expected behavior in this specific example is that the connection is made without an error, the queue icon should be shown with the queue listed underneath and the topic icon should also be shown, but without any topics listed underneath.

**Comment by SeanFeldman** [comment]
Just to confirm, you're talking about scenarios where connection string looks like the following: `Endpoint=sb://...;EntityPath=topicname`?

If that's the case, yes, it does fail today. SBExplorer is a management tool and by default it expects to be able to iterate over entities. I think the issue is more of a feature request. Something along the lines of "Enable support for a single entity connection string".

**Comment by ErikMogensen** [comment]
It could for a start just parse the connection string and if ends with ;entityPath=... show a messagebox saying that such connection string are currently not supported.

**Comment by SeanFeldman** [comment]
> It could for a start just parse the connection string and if ends with ;entityPath=... show a messagebox saying that such connection string are currently not supported.

@ErikMogensen I'd like to understand the feature requirement prior to suggesting any solution 🙂
Perhaps our friend @atyourservers will be willing to submit a PR that unblocks this functionality. Who knows. The first step is to label the feature correctly.

**Comment by SeanFeldman** [comment]
Ironically, last Friday I ran into this issue where customers tried to use SBE with an entity-specific connection string 😀

**Comment by atyourservers** [comment]
@SeanFeldman, yes, users are trying to connect to an entity directly. I need to test again what happens when connecting to an entity, say a queue, when only that entity type is selected. I believe that worked without displaying an error, so if the same is true for topics, then instead of a feature request, it's more a matter of educating our users to select only that entity type that is relevant for the connection string. ;)

**Comment by paolosalvatori** [comment]
Entity specific connection strings are not supported as @SeanFeldman said. The tool requires a namespace-level connection string with the manage claim.

**Comment by SeanFeldman** [comment]
I've renamed the issue to represent the feature @atyourservers has described.
A separate issue (https://github.com/paolosalvatori/ServiceBusExplorer/issues/376) raised to fix the current behaviour.

**Comment by darrenshrwd** [comment]
In our enterprise we plan to use the same servicebus for many teams, however we would only want individual teams to have access to manage their specific topics or queues. So this feature would be extremely useful to us...as such I may well be willing to contribute the code and start next week during our 'hackathon' event.  If someone could give me pointers to get started between now and then that would be great.

**Comment by paolosalvatori** [comment]
@darrenshrwd do you have bandwidth to work on this feature and submit a PR?

**Comment by darrenshrwd** [comment]
@paolosalvatori Unfortunately, given people can just get by with old releases (or indeed the current release when I last looked) I realised I couldn't really justify any time on it.

**Comment by rajrao** [comment]
What do you guys think about fixing this by parsing the connString and if it has an entityPath then extract it and set it as the topic and queue filter paths?

**Comment by paolosalvatori** [comment]
Hi @rajrao we could do that, but the connection string should contain the management claim for the entity in question. Now, the code still uses the NamespaceManager from the old client Microsoft.ServiceBus client SDK, so I don't know if this enough and allowed, we should check! So you the bandwidth to give it a try? :)

**Comment by rajrao** [comment]
This looks like its already working correctly. I had to remove the check on the entityName check and also fix an issue with FilterOutException (which must have happened when the code was changed to be async). After that it works! The only thing is that it spits out an ugly error as it spits out the whole stack (not just the message). Thought of changing that, but the entire stack can be nice!

Tested both topic and queue based connection strings
Endpoint=sb://xxxx.servicebus.windows.net/;SharedAccessKeyName=test;SharedAccessKey=abcd;EntityPath=topicPath
